### PR TITLE
MPLUGINTESTING-93 Documentation for Plugin testing

### PR DESCRIPTION
**Pull request by aamotharald**
Added Docu on how to test with JUnit5.
Converted the Docu from apt to md .
This PR should only get merged once org.apache.maven.plugin-testing:maven-plugin-testing-harness:3.4.0 has been released as on 3.3.0 there is no JUnit5 testing support at all.

### CODE REVIEW

File: content/markdown/plugin-developers/plugin-testing.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+A functional/integration test attempts to use a mojo in a real Maven build, by launching a real instance of Maven in a real project. Normally this requires you to construct special dummy Maven projects with real POM files. Often this requires you to have already installed your plugin into your local repository so it can be used in a real Maven build. Functional tests run much more slowly than unit tests, but they can catch bugs that you may not catch with unit tests.
+
+The general wisdom is that your code should be mostly tested with unit tests, but should also have some functional tests.
+
+## Unit Tests
+
+### Using JUnit alone
+
+In principle, you can write a unit test of a plugin Mojo the same way you'd write any other JUnit test case, by writing a class that `extends TestCase`.
```
Comment (slawekjaranowski) [inline content/markdown/plugin-developers/plugin-testing.md:37]: `TestCase` is from JUnit 3 I would like drop such sentence at all

File: content/markdown/plugin-developers/plugin-testing.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+      <scope>test</scope>
+    </dependency>
+    ...
+  </dependencies>
+...
+```
+
+#### JUnit5 style tests
+
+JUnit5 (jupiter) uses an Extension framework for which the `MojoExtension` is provided by the `maven-plugin-testing-harness`.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:72]: The JUnit 5 User Guide always writes `JUnit Jupiter` (as it contains the extension programming model) so I suggest to do this too here and everywhere.
In the heading I suggest to mention JUnit 5 once, e.g. `JUnit Jupiter (JUnit 5) style tests`.

P.S. `extension framework` in lowercase.

File: content/markdown/plugin-developers/plugin-testing.md (hunk 1 above)
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:37]: As @slawekjaranowski already mentioned, I also suggest to delete the `, by writing a class that..` part, because JUnit 4 and 5 provide other ways and those depend on the version you use - and those have own sections later.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+The general wisdom is that your code should be mostly tested with unit tests, but should also have some functional tests.
+
+## Unit Tests
+
+### Using JUnit alone
+
+In principle, you can write a unit test of a plugin Mojo the same way you'd write any other JUnit test case, by writing a class that `extends TestCase`.
+
+However, many mojo methods need more information to work properly. For example, you&apos;ll probably need to inject a reference to a `MavenProject`, so your mojo can query project variables.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:39]: Yo don't need to write `you&apos;ll`, just write `you'll` or spend time to write the two missing characters to `you probably will need to`

File: content/markdown/plugin-developers/plugin-testing.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+### Using PlexusTestCase
+
+Mojo variables are injected by Guice, sometimes with a Plexus adapter to support the legacy `@Component` annotation. Currently some mojos are fully guicified with constructor injection, while others that have not yet been converted use Plexus field injection.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:43]: What are `Guice` and `Plexus`? You should give a short information.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+Both Guice-based and Plexus-based mojos rely on the Guice Plexus adapter to inject dependencies by having the test class extend `PlexusTestCase` and calling the **lookup()** method to instantiate the mojo. Tests for fully Guicified mojos can also inject dependencies directly into the constructor without extending `PlexusTestCase`. These dependencies can be Mockito mocks or instances of the actual model classes. If a particular test does not access the injected field — that is, it&apos;s only injected to fulfill the constructor signature — you can usually also pass null as the value of that argument. 
+
+With that said, if you need to inject Maven objects into your mojo, you&apos;ll probably prefer to use the maven-plugin-testing-harness.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:47]: see above about `&apos;`

File: content/markdown/plugin-developers/plugin-testing.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+### Using the maven-plugin-testing-harness
+
+The [maven-plugin-testing-harness](/plugin-testing/maven-plugin-testing-harness/) is explicitly intended to test the `org.apache.maven.reporting.AbstractMavenReport#execute()` implementation.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:51]: I miss an object behind the "implementation", like "implementation of the plugin" or something. Reads a bit smoother. But that may be personal.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+In general, you need to include `maven-plugin-testing-harness` as a test-scoped dependency.
+
+```xml
+...
+  <dependencies>
+    ...
+    <dependency>
+      <groupId>org.apache.maven.plugin-testing</groupId>
+      <artifactId>maven-plugin-testing-harness</artifactId>
+      <version>3.4.0</version>
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:62]: Would add a comment (preferred in the XML section 'cause of "copy and paste without reading people", that the current version (and not fixed 3.4.0) should be used.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    @Test
+    @InjectMojo(goal = "generate", pom = POM)
+    void simpleMojo(YourMojo mojo) {
+        assertNotNull( mojo );
+    }
+}
+```
+
+#### deprecated JUnit4 style tests
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:91]: To align this with the JUnit Jupiter section
`JUnit 4 style tests (deprecated)`

File: content/markdown/plugin-developers/plugin-testing.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+There is the deprecated way to write tests using JUnit4 style.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:92]: Suggestion:

> For Maven 3 only you can still write tests, using the JUnit 4 style. But writing tests this way is deprecated and won't work for Maven 4 anymore, because Maven 4 supports only JUnit Jupiter!

File: content/markdown/plugin-developers/plugin-testing.md [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+     * @see junit.framework.TestCase#setUp()
+     */
+    protected void setUp() throws Exception
+    {
+        // required for mojo lookups to work
+        super.setUp();
+    }
+
+    /**
+     * @throws Exception
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:112]: The method signature already points out that `Exception` may get thrown.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+          "src/test/resources/unit/basic-test/basic-test-plugin-config.xml" );
+
+        YourMojo mojo = (YourMojo) lookupMojo( "yourGoal", testPom );
+
+        assertNotNull( mojo );
+    }
+}
+```
+
+#### more information
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:126]:
```
#### More information
For more information, please refer to the [Maven Plugin Harness Wiki](https://cwiki.apache.org/confluence/display/MAVENOLD/Maven+Plugin+Harness)
```

File: content/markdown/plugin-developers/plugin-testing.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+For more information, refer to [Maven Plugin Harness Wiki](https://cwiki.apache.org/confluence/display/MAVENOLD/Maven+Plugin+Harness)
+
+## Integration/Functional testing
+
+### maven-verifier
+
+maven-verifier tests are run using JUnit, and provide a simple class allowing you to launch Maven and assert on its log file and built artifacts. It also provides a `ResourceExtractor`, which extracts a Maven project from the src/test/resources directory into a temporary working directory where you can do tricky stuff with it. Follow the [Getting Started](/shared/maven-verifier/getting-started.html) guide to learn more about creating maven-verifier tests.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:133]: This sentence is hard to understand (and contains a lot of "and"). Can you rephrase it?

File: content/markdown/plugin-developers/plugin-testing.md [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+Maven itself uses maven-verifier to run its core integration tests. For more information, see [Creating a Maven Integration Test](https://cwiki.apache.org/confluence/display/MAVEN/Creating+a+Maven+Integration+Test).
+
+**Note**: maven-verifier and maven-verifier-plugin sound similar, but are totally different unrelated pieces of code. maven-verifier-plugin simply verifies the existence/absence of files on the filesystem. You could use it for functional testing, but you may need more features than maven-verifier-plugin provides.
+
+### maven-invoker-plugin
+
+You can use [maven-invoker-plugin](https://maven.apache.org/plugins/maven-invoker-plugin/) to invoke Maven and to provide some BeanShell/Groovy tests. Tests written in this way don&apos;t run under JUnit/TestNG; instead, they&apos;re run by Maven itself.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:141]: `%apos;`

File: content/markdown/plugin-developers/plugin-testing.md [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+You can take a look at the [maven-install-plugin](https://svn.apache.org/repos/asf/maven/plugins/trunk/maven-install-plugin/src/it/) to see how integration tests are written.
```
Comment (Bukama) [inline content/markdown/plugin-developers/plugin-testing.md:143]: Link to SVN?

**Review by Bukama — COMMENTED** [review]
Thanks for the migration. Got some comments, you may want to consider.

P.S. I personally prefer to write only one sentence per line in a documentation as code file, but I don't know if maven has clear rules about this.

File: content/markdown/plugin-developers/plugin-testing.md (hunk 3 above)
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:39]: can we use a proper curly apostrophe instead of &apos;?

File: content/markdown/plugin-developers/plugin-testing.md [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    ...
+  </dependencies>
+...
+```
+
+#### JUnit5 style tests
+
+JUnit5 (jupiter) uses an Extension framework for which the `MojoExtension` is provided by the `maven-plugin-testing-harness`. 
+You can annotate your JUnit5 test with `@MojoTest` and with that leverage the `MojoExtension` to inject the Mojo under test.
+This functionality got introduced with version `3.4.0` of the `maven-plugin-testing-harness`.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:74]: got --> was
with version --> in version

File: content/markdown/plugin-developers/plugin-testing.md [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+    void simpleMojo(YourMojo mojo) {
+        assertNotNull( mojo );
+    }
+}
+```
+
+#### deprecated JUnit4 style tests
+There is the deprecated way to write tests using JUnit4 style. 
+This is not recommended, but you can still use it on Maven 3. 
+For Maven 4 only JUnit5 style tests are available and JUnit4 will not be supported there anymore.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:94]: will not be supported there anymore --> is not supported

**Comment by aamotharald** [comment]
Thanks @slawekjaranowski @elharo and @Bukama for your review! I learned a lot and was able to address your review comments with (for me) sound solutions.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+This document is intended to help Maven Developers test plugins with unit tests, integration tests, and functional tests.
+
+## Testing Styles: Unit Testing vs. Functional/Integration Testing
+
+A unit test attempts to verify a mojo as an isolated unit, by mocking out the rest of the Maven environment.
+A mojo unit test does not attempt to run your plugin in the context of a real Maven build. Unit tests are designed to be fast.
+
+A functional/integration test attempts to use a mojo in a real Maven build, by launching a real instance of Maven in a real project.
+Normally this requires you to construct special dummy Maven projects with real POM files.
+Often this requires you to have already installed your plugin into your local repository so it can be used in a real Maven build.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:33]: delete this sentence. It's not true

File: content/markdown/plugin-developers/plugin-testing.md [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+In principle, you can write a unit test of a plugin Mojo the same way you’d write any other JUnit test case.
+
+However, many mojo methods need more information to work properly.
+For example, you’ll probably need to inject or mock a reference to a `MavenProject`, so your mojo can query project variables.
+
+### Using PlexusTestCase
+
+Mojo variables are injected by Guice (an open-source software framework for the Java platform), sometimes with a Codehaus Plexus (a collection of components used by Apache Maven) adapter to support the legacy `@Component` annotation. 
+Currently some mojos are fully guicified with constructor injection, while others that have not yet been converted use Plexus field injection.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:50]: delete this sentence

File: content/markdown/plugin-developers/plugin-testing.md [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+
+Both Guice-based and Plexus-based mojos rely on the Guice Plexus adapter to inject dependencies by having the test class extend `PlexusTestCase` and calling the **lookup()** method to instantiate the mojo.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:52]: No longer true. Some construct the object directly.
Also at head there are no plexus based mojos in maven core

File: content/markdown/plugin-developers/plugin-testing.md [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+Tests for fully Guicified mojos can also inject dependencies directly into the constructor without extending `PlexusTestCase`.
+These dependencies can be Mockito mocks or instances of the actual model classes.
+If a particular test does not access the injected field — that is, it’s only injected to fulfill the constructor signature — you can usually also pass null as the value of that argument. 
+
+With that said, if you need to inject Maven objects into your mojo, you’ll probably prefer to use the maven-plugin-testing-harness.
+
+### Using the maven-plugin-testing-harness
+
+The [maven-plugin-testing-harness](/plugin-testing/maven-plugin-testing-harness/) is explicitly intended to test the implementation of the `org.apache.maven.reporting.AbstractMavenReport#execute()` method.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:61]: intended --> designed

File: content/markdown/plugin-developers/plugin-testing.md [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+In general, you need to include `maven-plugin-testing-harness` as a test-scoped dependency.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:63]: delete "In general,"

File: content/markdown/plugin-developers/plugin-testing.md [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+      <scope>test</scope>
+    </dependency>
+    ...
+  </dependencies>
+...
+```
+
+#### JUnit Jupiter (JUnit 5) style tests
+
+JUnit Jupiter uses an extension framework for which the `MojoExtension` is provided by the `maven-plugin-testing-harness`.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:83]: delete "the"

File: content/markdown/plugin-developers/plugin-testing.md [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+    </dependency>
+    ...
+  </dependencies>
+...
+```
+
+#### JUnit Jupiter (JUnit 5) style tests
+
+JUnit Jupiter uses an extension framework for which the `MojoExtension` is provided by the `maven-plugin-testing-harness`. 
+You can annotate your JUnit Jupiter test with `@MojoTest` and with that leverage the `MojoExtension` to inject the Mojo under test.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:84]: "and with that leverage the `MojoExtension` to inject" -->
"; then inject the mojo under test into the test method with the @InjectMojo annotation"

File: content/markdown/plugin-developers/plugin-testing.md [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+```java
+@MojoTest
+public class YourMojoTest {
+
+    private static final String POM = "src/test/resources/unit/basic-test/basic-test-plugin-config.xml";
+
+    @Test
+    @InjectMojo(goal = "generate", pom = POM)
+    void simpleMojo(YourMojo mojo) {
+        assertNotNull( mojo );
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:97]: no spaces around parentheses

File: content/markdown/plugin-developers/plugin-testing.md [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+    }
+}
+```
+
+#### JUnit 4 style tests (deprecated)
+There is the deprecated way to write tests using JUnit 4 style. 
+This is not recommended, but you can still use it on Maven 3. 
+For Maven 4 only JUnit Jupiter style tests are available and JUnit 4 is not supported there anymore.
+Please consider migrating your JUnit 4 style MojoTests to JUnit Jupiter tests.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:106]: delete Please

File: content/markdown/plugin-developers/plugin-testing.md [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+    void simpleMojo(YourMojo mojo) {
+        assertNotNull( mojo );
+    }
+}
+```
+
+#### JUnit 4 style tests (deprecated)
+There is the deprecated way to write tests using JUnit 4 style. 
+This is not recommended, but you can still use it on Maven 3. 
+For Maven 4 only JUnit Jupiter style tests are available and JUnit 4 is not supported there anymore.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:105]: available. JUnit 4 is no longer supported.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+public class YourMojoTest
+    extends AbstractMojoTestCase
+{
+    /**
+     * @see junit.framework.TestCase#setUp()
+     */
+    protected void setUp() throws Exception
+    {
+        // required for mojo lookups to work
+        super.setUp();
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:119]: no it isn't. Delete this method

File: content/markdown/plugin-developers/plugin-testing.md [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    public void testMojoGoal() throws Exception
+    {
+        File testPom = new File( getBasedir(),
+          "src/test/resources/unit/basic-test/basic-test-plugin-config.xml" );
+
+        YourMojo mojo = (YourMojo) lookupMojo( "yourGoal", testPom );
+
+        assertNotNull( mojo );
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:129]: no spaces on parentheses

File: content/markdown/plugin-developers/plugin-testing.md [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+    }
+}
+```
+
+#### More information
+For more information, please refer to the [Maven Plugin Harness Wiki](https://cwiki.apache.org/confluence/display/MAVENOLD/Maven+Plugin+Harness)
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:135]: delete this. The wiki is not remotely up to date

File: content/markdown/plugin-developers/plugin-testing.md [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+
+## Integration/Functional testing
+
+### maven-verifier
+
+maven-verifier tests are run using JUnit.
+They provide a simple class allowing you to launch Maven and assert on its log file and its build artifacts.
+It also provides a `ResourceExtractor`, which extracts a Maven project from the src/test/resources directory into a temporary working directory where you can do tricky stuff with it.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:143]: They or It?

**Comment by Bukama** [comment]
Can't add anything to what @elharo has already mentioned.

File: content/markdown/plugin-developers/plugin-testing.md (hunk 19 above)
Comment (aamotharald) [reply content/markdown/plugin-developers/plugin-testing.md:47]: I have marked the JUnit 4 style tests as deprecated without any more Docu on how to create such tests. How is the situation for PlexusTestCase. Does it make sense to mark that as deprecated analogously or does this keep getting maintained?

File: content/markdown/plugin-developers/plugin-testing.md [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+
+The general wisdom is that your code should be mostly tested with unit tests, but should also have some functional tests.
+
+## Unit Tests
+
+### Using JUnit alone
+
+In principle, you can write a unit test of a plugin Mojo the same way you’d write any other JUnit test case.
+
+However, many mojo methods need more dependencies to work properly.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:40]: Is it Mojo or mojo? I think mojo though I'm not sure, but either way let's be consistent.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 32]
```diff
@@ -0,0 +1,154 @@
...
+For example, you’ll probably need to inject or mock a reference to a `MavenProject`, so your mojo can query project variables.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:41]: delete "or mock" as mocks still need to be injected.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 33]
```diff
@@ -0,0 +1,154 @@
...
+
+### Using PlexusTestCase
+
+Mojo variables are injected by Guice (an open-source software framework for the Java platform), sometimes with a Codehaus Plexus (a collection of components used by Apache Maven) adapter.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:45]: delete "(an open-source software framework for the Java platform)" as it doesn't add much. Folks can Google it if they've never heard of it

File: content/markdown/plugin-developers/plugin-testing.md [hunk 34]
```diff
@@ -0,0 +1,154 @@
...
+In principle, you can write a unit test of a plugin Mojo the same way you’d write any other JUnit test case.
+
+However, many mojo methods need more dependencies to work properly.
+For example, you’ll probably need to inject or mock a reference to a `MavenProject`, so your mojo can query project variables.
+
+### Using PlexusTestCase
+
+Mojo variables are injected by Guice (an open-source software framework for the Java platform), sometimes with a Codehaus Plexus (a collection of components used by Apache Maven) adapter. 
+
+Both Guice-based and Plexus-based mojos rely on the Guice Plexus adapter to inject dependencies by having the test class extend `PlexusTestCase` and calling the **lookup()** method to instantiate the mojo.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:47]: delete "rely on the Guice Plexus adapter to" as Plexus mojos don't do this.

File: content/markdown/plugin-developers/plugin-testing.md (hunk 19 above)
Comment (elharo) [reply content/markdown/plugin-developers/plugin-testing.md:47]: PlexusTestCase is in a different project and repo, so can't be deprecated here. I would like to get rid of it, but I'm not sure I can. It is overused.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 35]
```diff
@@ -0,0 +1,154 @@
...
+Tests for fully guicified mojos can also inject dependencies directly into the constructor without extending `PlexusTestCase`.
+These dependencies can be Mockito mocks or instances of the actual model classes.
+If a particular test does not access the injected field — that is, it’s only injected to fulfill the constructor signature — you can usually also pass null as the value of that argument.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:50]: delete "also"

File: content/markdown/plugin-developers/plugin-testing.md [hunk 36]
```diff
@@ -0,0 +1,154 @@
...
+You have to include `maven-plugin-testing-harness` as a test-scoped dependency.
+
+```xml
+...
+  <dependencies>
+    ...
+    <dependency>
+      <groupId>org.apache.maven.plugin-testing</groupId>
+      <artifactId>maven-plugin-testing-harness</artifactId>
+      <!-- use the latest version of the maven-plugin-testing-harness, >= 3.4.0 -->
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:67]: delete comment

File: content/markdown/plugin-developers/plugin-testing.md [hunk 37]
```diff
@@ -0,0 +1,154 @@
...
+JUnit Jupiter uses an extension framework for which `MojoExtension` is provided by the `maven-plugin-testing-harness`. 
+You can annotate your JUnit Jupiter test with `@MojoTest`; then inject the mojo under test into the test method with the `@InjectMojo` annotation.
+This functionality was introduced in version `3.4.0` of the `maven-plugin-testing-harness`.
+Below is an example:
+
+```java
+@MojoTest
+public class YourMojoTest {
+
+    private static final String POM = "src/test/resources/unit/basic-test/basic-test-plugin-config.xml";
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:87]: I'm tempted to inline this. Named constants for single use strings aren't helpful.

File: content/markdown/plugin-developers/plugin-testing.md [hunk 38]
```diff
@@ -0,0 +1,154 @@
...
+    @Test
+    @InjectMojo(goal = "generate", pom = POM)
+    void simpleMojo(YourMojo mojo) {
+        assertNotNull(mojo);
+    }
+}
+```
+
+#### JUnit 4 style tests (deprecated)
+For Maven 3 there is the deprecated way to write tests using JUnit 4 style.
```
Comment (elharo) [inline content/markdown/plugin-developers/plugin-testing.md:98]: Rewrite this paragraph and don't mark it deprecated. Maybe, Maven 3 allows JUnit 4 style tests. Maven 4 only allows Junit Jupiter tests.

File: content/markdown/plugin-developers/plugin-testing.md (hunk 33 above)
Comment (elharo) [reply content/markdown/plugin-developers/plugin-testing.md:45]: and might as wellm delete "(a collection of components used by Apache Maven)" too

**Comment by slawekjaranowski** [comment]
Thanks for work ... By the way I would like to have such documentation in project in branch 3.x
https://github.com/apache/maven-plugin-testing/tree/maven-plugin-testing-3.x

I'm still think that release of 3.4.0 - maven-plugin-testing should be done ... on my list 😄

**Review by Bukama — APPROVED** [review]

**Comment by Bukama** [comment]
@aamotharald Sorry for the long time not responding. We are still interested in the changes of your PR. Would you refresh it (put in last comments and resolve conflicts) :)

**Comment by slawekjaranowski** [comment]
@aamotharald thanks for the effort, we have done similar updates in:
 - https://github.com/apache/maven-site/pull/1435
 - https://github.com/apache/maven-site/pull/1437

so I close it
